client/state/document-head/middleware.js

```diff
@@ -2,27 +2,80 @@
  * Internal dependencies
  */
 import { DOCUMENT_HEAD_TITLE_SET, DOCUMENT_HEAD_UNREAD_COUNT_SET } from 'state/action-types';
-import { setTitle as legacySetTitle, setCount as legacySetCount } from 'lib/screen-title/actions';
-import { getSelectedSiteId } from 'state/ui/selectors';
+import { decodeEntities } from 'lib/formatting';
+import { getTitle, getUnreadCount } from './selectors';
+import { getSelectedSite } from 'state/ui/selectors';
 
 /**
  * Middleware that updates the screen title when a title updating action is
- * dispatched. Currently, this dispatches through the legacy Flux actions used
- * elsewhere in the codebase.
+ * dispatched.
  *
  * @param {Object} store Redux store instance
  * @returns {Function} A configured middleware with store
  */
 export default ( { getState } ) => ( next ) => ( action ) => {
+	let formattedTitle;
 	switch ( action.type ) {
 		case DOCUMENT_HEAD_TITLE_SET:
-			legacySetTitle( action.title, { siteID: getSelectedSiteId( getState() ) } );
+			formattedTitle = getFormattedTitle(
+				action.title,
+				getUnreadCount( getState() ),
+				getSelectedSite( getState() )
+			);
+			if ( formattedTitle !== document.title ) {
+				document.title = formattedTitle;
+			}
 			break;
 
 		case DOCUMENT_HEAD_UNREAD_COUNT_SET:
-			legacySetCount( action.count, { siteID: getSelectedSiteId( getState() ) } );
+			formattedTitle = getFormattedTitle(
+				getTitle( getState() ),
+				action.count,
+				getSelectedSite( getState() )
+			);
+			if ( formattedTitle !== document.title ) {
+				document.title = formattedTitle;
+			}
 			break;
 	}
 
 	return next( action );
 };
+
+function getFormattedTitle( title, unreadCount, site ) {
```

Comment:
> [...] Maybe a [store enhancer](http://redux.js.org/docs/Glossary.html#store-enhancer)?

I'll look that tomorrow. Other than that, I think this PR is ready for review, and since that will probably take a while, I'm setting it to 'Needs Review' already.